### PR TITLE
Feature: Multiple Timelines

### PR DESCRIPTION
**Pull request by kfriars**
## New Feature
Allow the content to be viewed in multiple timelines.

### Summary
By adding a new model/table and scoping the queries against the timelines, you can provide concurrent but unique views to the content in your database.

### Tasks
1. Create timeline Tables + append to existing migrations
2. Add Timeline Model and create relevant relations
3. Add Timelines to the RevisionableScope queries
4. Add the notion of an active Checkpoint for querying / revisioning content
5. Update Readme
6. Add some Feature Tests
7. Add psalm coverage

### CODE REVIEW

File: src/Models/Timeline.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+use Illuminate\Database\Eloquent\Collection;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\HasMany;
+
+/**
+ * @property-read Collection|Checkpoint[] $checkpoints
+ */
+class Timeline extends Model
+{
+    /**
```
Comment (m-triassi) [inline src/Models/Timeline.php:14]: I believe timelines are missing a `$fillable` property? unless i'm missing some other avenue to create timelines?

Something like this?
 ```suggestion
class Timeline extends Model
{

    protected $fillable = ['title']
    
    /**
```

File: src/Models/Timeline.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+     * @return HasMany
+     */
+    public function checkpoints(): HasMany
+    {
+        /** @var Checkpoint|string $checkpointClass */
+        $checkpointClass = config('checkpoint.models.checkpoint');
+
+        return $this->hasMany($checkpointClass, $checkpointClass::TIMELINE_ID);
+    }
+}
```
Comment (m-triassi) [inline src/Models/Timeline.php:26]: Are we also maybe missing a `setActive()` on timelines as well? i see that works pretty well with checkpoints. In this use case are users just expected to filter on timelines themselves?

**Review by m-triassi — CHANGES_REQUESTED** [review]
Couple of comments here; Also looks like Psalm is failing, let's fix that as well? 

as an aside i wonder if we need more imposed structure here with the introduction of timelines as far as i understand, a time line is to allow concurrent checkpoints chains, right? so in that case we might want to allow swapping which timeline in active, and automatically filter by that? 

Or maybe the intention is that a user will swap to some checkpoint on a different time line and that implies the timeline we're on?

Let me know if i'm understanding here :) maybe we can rationalize the addition in the docs more?

File: src/Stores/BasicCheckpointStore.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+use Plank\Checkpoint\Models\Checkpoint;
+
+class BasicCheckpointStore implements CheckpointStore
+{
+    /**
+     * A variable to store the active Checkpoint
+     *
+     * @var null|Checkpoint
+     */
+    protected ?Checkpoint $checkpoint = null;
```
Comment (kfriars) [inline src/Stores/BasicCheckpointStore.php:15]: Fix for php 7.

File: src/Scopes/RevisionScope.php [hunk 4]
```diff
@@ -60,21 +66,30 @@ public function extend(Builder $builder)
...
-            $builder->whereHas('revision', function (Builder $query) use ($model, $until, $since) {
-                $query->whereIn($query->getModel()->getQualifiedKeyName(),
-                    $query->newModelInstance()->setTable('_r')->from($query->getModel()->getTable(), '_r')
-                        ->latestIds($until, $since)
-                        ->whereColumn($query->qualifyColumn('original_revisionable_id'), '_r.original_revisionable_id')
-                        ->whereType($model));
+            $builder->whereHas('revision', function (Builder $query) use ($model, $until, $since, $timeline, $revision) {
+                $timelineId = $timeline ? $timeline->getKey() : null;
+
+                $query->where($revision::TIMELINE_ID, $timelineId)
```
Comment (kfriars) [inline src/Scopes/RevisionScope.php:84]: Remove unnecessary timeline scope.

File: src/Stores/BasicCheckpointStore.php (hunk 3 above)
Comment (a-drew) [reply src/Stores/BasicCheckpointStore.php:15]: good catch, I haven't spotted all of them.

**Review by a-drew — APPROVED** [review]
Good to go 🚀

**Review by m-triassi — APPROVED** [review]
Kurt dealt with all my concerns here! nice work you two! :)

**Pull request merged**